keps/sig-node/2000-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,843 @@
+critical work prior to shutdown/sleep.
+
+We believe that making use of systemd is a reasonable approach considering
+almost all new popular linux distros are systemd based (RHEL, Google COS,
+Ubuntu, CentOS, Debian, Fedora, Flatcar Linux, see widespread
+[adoption](https://en.wikipedia.org/wiki/Systemd#Adoption)) and [systemd
+183](https://lwn.net/Articles/499480/) (released in 2012) features support for
+inhibitors.
+
+Thanks to @giuseppe for helping with getting systemd inhibitors working!
+
+### Implementation
+
+Introduce a new Kubelet Config setting, `kubeletConfig.ShutdownGracePeriod`,
+defaulting to 0 seconds. Upon kubelet startup,
+
+*   if the setting is greater than 0 seconds
+    *   kubelet will check with dbus current `InhibitDelayMaxSec` to check if
+        `kubeletConfig.ShutdownGracePeriod &lt;=` `InhibitDelayMaxSec`.
+*   if `kubeletConfig.ShutdownGracePeriod` > `InhibitDelayMaxSec`
+    *   Kubelet will attempt to update the InhibitDelayMaxSec setting, by
+        writing a config file to `/etc/systemd/logind.conf.d/kubelet.conf`, and
+        sending a SIGHUP to logind to update the config setting to ensure that
+        the ShutdownGracePeriod from kubelet config is equal to
+        `InhibitDelayMaxSec`.
+
+After updating the `InhibitDelayMaxSec` on the node if needed, Kubelet will
+query the dbus for the final value of `InhibitDelayMaxSec` set on the node and
+treat min(`InhibitDelayMaxSec`, `kubeletConfig.ShutdownGracePeriod`) as the
+allocatable  shutdown grace period, which will be referred to in this KEP as
+`ShutdownGracePeriod`.
+
+Kubelet will register with dbus as a delay systemd inhibitor lock for the
+`ShutdownGracePeriod` for the shutdown event. Kubelet will also register a
+`PrepareForShutdown` signal which will be emitted prior to the shutdown. Upon
+receiving the signal, Kubelet will have additional `ShutdownGracePeriod` time
+before the actual node will initiate the shutdown.
+
+**Handling the shutdown**
+
+Upon a shutdown occurring, Kubelet will gracefully terminate all the pods
+running on the node and taint itself with a new taint,
+`node.kubernetes.io/unschedulable` to avoid new pods being scheduled.
+
+Since some of the pods running on the node are often critical for the the
+workloads running on a node (e.g. logging pod daemonset, kubeproxy, kubedns)
+etc, we choose to split the pods running on the node into two categories,
+“critical system pods”, and regular pods. Critical system pods should be
+terminated last, because for example, if the logging pod is terminated first,
+logs from the other workloads will not be captured. Critical system pods are
+identified as those that are in the `system-cluster-critical` or
+`system-node-critical` [priority
+classes](https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/#marking-pod-as-critical).
+
+Upon shutdown Kubelet will:
+
+1. Gracefully terminate all non critical system pods with a gracePeriodOverride
```

Comment:
do we need a buffer to allow the kubelet to report back to the apiserver?

Reply:
to clarify, are you proposing before pods are terminated we first report status to apiserver, and only then start terminating pods?

Reply:
I also hope we can have the status back to the apiserver, but hope the status include all pods are terminated or being terminated. Not sure if we can guarantee it though.

Reply:
I updated this to include that the upon shutdown, kubelet should also update it's Ready status to false with the reason `Node is shutting down`. 

Regarding the pod status being reported back as well, I think this would happen as part of the regular sync loop, but it might be difficult to guarantee depending on the timing as Dawn mentioned.